src/pybind/mgr/dashboard/controllers/host.py

```diff
@@ -272,15 +278,23 @@ class Host(RESTController):
     @EndpointDoc("List Host Specifications",
                  parameters={
                      'sources': (str, 'Host Sources'),
+                     'facts': (bool, 'Host Facts')
                  },
                  responses={200: LIST_HOST_SCHEMA})
-    def list(self, sources=None):
-        if sources is None:
-            return get_hosts()
-        _sources = sources.split(',')
-        from_ceph = 'ceph' in _sources
-        from_orchestrator = 'orchestrator' in _sources
-        return get_hosts(from_ceph, from_orchestrator)
+    @RESTController.MethodMap(version='1.1')
+    def list(self, sources=None, facts=False):
+        hosts = get_hosts(sources)
+        orch = OrchClient.instance()
+        if str_to_bool(facts):
+            if orch.available():
+                hosts_facts = orch.hosts.get_facts()
+                return merge_list_of_dicts_by_key(hosts, hosts_facts, 'hostname')
```

Comment:
The fact that we need this `merge_list_of_dicts_by_key()` weird function is a good smell of something not very healthy. I think we shouldn't make our back-end methods just by-passers of data. We should always gather the data we need from the inner/deeper layers, compose our response with our own format and only return the minimal set of required info, and that alone...

I know this has not started with this PR, it's a long tradition of letting data pass unprocessed through our back-end (which IMHO has always been a bad practice, as we don't have explicit data models of the info we deal with).

So, how can we commit to version the API if we're the owners of the data model we're returning... We'll always be bound to whatever changes other components introduce.

Why not creating an explicit model of the data you're returning here? We're actually validating that implicit model in the API tests... but surprisingly that model is not in our production code, only in test code... (I think these are partly the perils of type-less languages).

Reply:
Yes, as discussed I'm trying to  design model for `HostList`, but while doing it there are many cleanups I have to do first, so I was thinking how about I deal with all these cleanups [here](https://github.com/ceph/ceph/pull/41959) itself and move forward with this PR without introducing any more changes?

Reply:
Can I help you with the model here?

This could be the model:
```py
class HostFacts(NamedTuple):
  IPv4_addresses: List[IPv4Address]
  IPv6_addresses: List[IPv6Address]
  ...
  memory_total_bytes: int
  hdd_capacity_bytes:int
  flash_capacity_bytes: int
  raw_capacity: int

  
class Host(NamedTuple):
  name: str
  labels: List[str]
  ...
  facts: HostFacts
```

What we want here is something like this:
```suggestion
                return [Host(hostname=host.hostname, ...) for host in get_hosts()]
```

But we have several challenges here:
* Instead of manually entering all the class init arguments, we could use the `Host(**host)` construct and filter out the arguments our class doesn't support. The `NamedTuple` doesn´t allow to override the __init__ method, but we can create a factory method (@classmethod):
```py
  @classmethod
  from_dict(cls, host, facts):
    return cls(facts=facts, **{key: value for key, value in host.items() if key in cls._fields})
```
* If we need to change field names, I'd rather create an external helper for renaming keys in dictionaries.
* Additionally, I'd check what model (if any) the `get_hosts()` uses and use that if possible. If we do this, we'll discover a `HostSpec` class in `src/python-common/ceph/deployment/`. We'll also find a `HostFacts` class under `src/pybind/mgr/cephadm/configchecks.py`.
* Another issue is that we need to populate data from 2 different sources `get_hosts` and `get_facts` and those are 2 lists, not dicts. You have used this `merge_list_of_dicts_by_key` and directly return what you get from there... I'd rather create an index (or dict) for the facts and that's all.

This would look like this:
```py
  facts_index = {facts['hostname']: fact for fact in facts}
  return [ost.from_dict(host, facts_index[host.hostname]) for host in get_hosts()]
```